index/full_doc_id_position_iterator.go

```diff
@@ -0,0 +1,51 @@
+package index
+
+// fullDocIDPositionIterator is a (doc ID, position) iterator for a full
+// doc ID set representing [0, numTotalDocs) masked by the given doc ID
+// set iterator `maskingIt`.
+type fullDocIDPositionIterator struct {
+	numTotalDocs int32
+	maskingIt    DocIDSetIterator
+
+	done         bool
+	currDocID    int32
+	currPosition int
+}
+
+func newFullDocIDPositionIterator(
+	numTotalDocs int32,
+	maskingIt DocIDSetIterator,
+) *fullDocIDPositionIterator {
+	return &fullDocIDPositionIterator{
+		numTotalDocs: numTotalDocs,
+		maskingIt:    maskingIt,
+		currDocID:    invalidDocID,
+		currPosition: -1,
+	}
+}
+
+func (it *fullDocIDPositionIterator) Next() bool {
+	if it.done {
+		return false
+	}
+	if !it.maskingIt.Next() {
+		it.done = true
+		return false
+	}
+	it.currPosition++
+	it.currDocID = it.maskingIt.DocID()
+	if it.currDocID < it.numTotalDocs {
```

Comment:
I might be totally misunderstanding something; I was under the assumption that doc ids are monotonically increasing, ie the DocIDSetIterator might start at 1000 and return 1000, 1001, 1002, etc. However here we're comparing against a `full doc ID set representing [0, numTotalDocs)` which can only occur naturally once given the doc ids are monotonically increasing.

Reply:
Hm not sure I get what you meant by "can only occur naturally once"?

Reply:
Would prefer to be a bit defensive here in case the masking iterator goes out of range.